pkg/discovery/worker/k8s_discovery_worker.go

```diff
@@ -0,0 +1,243 @@
+	if !exist {
+		configs = []monitoring.MonitorWorkerConfig{}
+	}
+
+	configs = append(configs, config)
+	c.monitoringSourceConfigs[monitorType] = configs
+
+	return c
+}
+
+// k8sDiscoveryWorker receives a discovery task from dispatcher(DiscoveryClient). Then ask available monitoring workers
+// to scrape metrics source and get topology information. Finally it builds entityDTOs and send back to DiscoveryClient.
+type k8sDiscoveryWorker struct {
+	// A UID of a discovery worker.
+	id string
+
+	// config
+	config *k8sDiscoveryWorkerConfig
+
+	// a collection of all monitoring worker of different types.
+	// key: monitoring types; value: monitoring worker instance.
+	monitoringWorker map[types.MonitorType][]monitoring.MonitoringWorker
+
+	// sink is a central place to store all the monitored data.
+	sink *metrics.EntityMetricSink
+
+	taskChan chan *task.Task
+}
+
+func NewK8sDiscoveryWorker(config *k8sDiscoveryWorkerConfig) (*k8sDiscoveryWorker, error) {
+	id := uuid.NewUUID().String()
+
+	if len(config.monitoringSourceConfigs) == 0 {
+		return nil, errors.New("No monitoring source config found in config.")
+	}
+
+	// Build all the monitoring worker based on configs.
+	monitoringWorkerMap := make(map[types.MonitorType][]monitoring.MonitoringWorker)
+	for monitorType, configList := range config.monitoringSourceConfigs {
+		monitorList, exist := monitoringWorkerMap[monitorType]
+		if !exist {
+			monitorList = []monitoring.MonitoringWorker{}
+		}
+		for _, config := range configList {
+			monitoringWorker, err := monitoring.BuildMonitorWorker(config.GetMonitoringSource(), config)
+			if err != nil {
+				// TODO return?
+				glog.Errorf("Failed to build monitoring worker configuration: %v", err)
+				continue
+			}
+			monitorList = append(monitorList, monitoringWorker)
+		}
+		monitoringWorkerMap[monitorType] = monitorList
+	}
+
+	return &k8sDiscoveryWorker{
+		id:               id,
+		config:           config,
+		monitoringWorker: monitoringWorkerMap,
+		sink:             metrics.NewEntityMetricSink(),
+
+		taskChan: make(chan *task.Task),
+	}, nil
+}
+
+func (worker *k8sDiscoveryWorker) RegisterAndRun(dispatcher *Dispatcher, collector *ResultCollector) {
+	worker.registerToDispatcher(dispatcher)
+	//worker.registerToResultCollector(collector)
+	for {
+		select {
+		case currTask := <-worker.taskChan:
+			glog.V(3).Infof("Worker %s has received a discovery task.", worker.id)
+			result := worker.executeTask(currTask)
+			collector.ResultPool() <- result
+			glog.V(3).Infof("Worker %s has finished the discovery task.", worker.id)
+
+			worker.registerToDispatcher(dispatcher)
+
+		}
+	}
+}
+
+func (worker *k8sDiscoveryWorker) registerToDispatcher(dispatcher *Dispatcher) {
+	glog.V(3).Infof("Register current worker %s to dispatcher", worker.id)
+	dispatcher.WorkerPool() <- worker.taskChan
+}
+
+// Worker start to working on task.
+func (worker *k8sDiscoveryWorker) executeTask(currTask *task.Task) *task.TaskResult {
```

Comment:
I think currently it is a two-stage operating, although the logic are written in the same function. I can break down the executeTask function into two sub-functions